source/common/upstream/upstream_impl.h

```diff
@@ -72,6 +72,42 @@ class HealthCheckHostMonitorNullImpl : public HealthCheckHostMonitor {
   void setUnhealthy(UnhealthyType) override {}
 };
 
+/**
+ * Implementation of LoadMetricStats.
+ *
+ * TODO(pianiststickman): this implementation puts a copy of the stat name into every host that
+ * receives a copy of that metric. This can be improved by putting a single copy of the stat name
+ * into a thread-local key->index map and using the index as the key to the stat map instead.
+ */
+class LoadMetricStatsImpl : public LoadMetricStats {
+public:
+  LoadMetricStatsImpl() = default;
+
+  void add(const absl::string_view key, double value) override {
+    absl::MutexLock lock(&mu_);
+    if (map_ == nullptr) {
+      map_ = std::make_unique<StatsMap>();
+    }
+    Stat& stat = (*map_)[key];
+    ++stat.num_requests_with_metric;
+    stat.total_metric_value += value;
+  }
+
+  StatsMapPtr latch() override {
+    absl::MutexLock lock(&mu_);
+    StatsMapPtr latched = std::move(map_);
+    map_ = nullptr;
+    return latched;
+  }
+
+private:
+  absl::Mutex mu_;
+  StatsMapPtr map_ ABSL_GUARDED_BY(mu_) = nullptr;
+
+  LoadMetricStatsImpl(const LoadMetricStatsImpl&) = delete;
```

Comment:
not strictly needed as you have 2 non-copyable members in here, but it's up to you. I suspect that if you do drop these you can also drop the default ctor.

Reply:
Done.